features/375-static-cdn/design.md

```diff
@@ -0,0 +1,142 @@
+- SPA style routing should be supported.
+
+## Design
+
+### Static Deployment
+
+A new type of deployment, `static`, would be supported. This deployment type
+supports configurating a deployment context (`context`), error page absolute
+path (`error_page`), and index page file name (`index_file`).
+
+When `static` deployment type is used:
+- The deployment context would be persisted to object storage
+    - symlinks are resolved at upload time
+    - `.skyignore` can be used to exclude files
+- A request with matching path would be routing using static deployment routing
+  algorithm (described below).
+- If a file is matched, the gateway would proxy the request to the backing
+  object storage file.
+- If no file is matched, a generic 404 page would be returned.
+- The response would have appropiate cache control headers, as configured by
+  developer.
+
+### Static Deployment Routing
+
+When a request matched a `static` deployment, only its path would be considered
+during routing. The request would be routed as followed:
+
+1. Let `U` be the request URI,
+   let `P` be the routing path without trailing slashes,
+   let `I` be the index page file name (default to `index.html`).
+2. If file `P` exists, return `P`.
+3. If file `Q` = `P/I` exists, return `Q`.
+4. If `error_page` is configured, set `U` = `U` with path `error_page` and route again.
+5. Otherwise, no file is matched.
+
+This procedure is similar to nginx directives:
+```
+index index.html
+try_files $uri <error_page>
+```
+
+Sometimes routing procedure is restarted (e.g. in step 4). The routing procedure
+can be performed at most 5 times to prevent infinite loop. In this case, `500
+Internal Server Error` response would be produced.
+
+**Example**:
+
+Suppose developer configured a `static` deployment as follow:
+```yaml
+  - name: assets
+    type: static
+    path: /
+    context: ./static
+    error_page: /index.html
+```
+
+- **https://example.com/main.js**:  
+    Route to `assets` and serve file at `./static/main.js`. (step 2)  
+- **https://example.com/**:  
+    Route to `assets` and serve file at `./static`. (step 2)  
+    Not a file, so instead serve file at `./static/index.html`. (step 3)  
+- **https://example.com/index.html**:  
+    Route to `assets` and serve file at `./static/index.html`. (step 2)  
+- **https://example.com/login**:  
+    Route to `assets` and serve file at `./static/login`. (step 2)  
+    File is not found, so instead serve file at `./static/login/index.html`. (step 3)  
+    File is not found, so instead rewrite URL with error page path
+    `https://example.com/index.html` and route again. (step 4)  
+    Route to `assets` and serve file at `./static/index.html`. (step 1)
+
+### Caching
+
+Developer can configure cache duration per `static` deployment. The gateway
+would attach appropiate headers to response according to configuration.
+
+Developers can disable caching by setting it explicitly to zero, but the
```

Comment:
> The unknown is a configuration with a default long values, someone will bombard us for support someday for this, I think we should either:

> 1. Make the default value short, maybe 10 mins as you proposed is good enough
> 2. Netlify's approach to this is it adds a fingerprint to all files, but I don't think we can replicate that...?
> 3. Make the value explicit
> 4. Enable cache invalidation feature

This sentence seems to me it means option 1. Am I correct?

Reply:
I intend it to be option 3, but I can see that the wordings can be confusing. I'll amend it further.